Refactor `ColumnDefinition` to contain `options` hash

kamipo opened the pull request:
Column options are passed as an hash args then used as `options` hash in
`add_column_options!`. Converting args to attributes is inconvinient for
using options as an hash.

kamipo committed:
Refactor `ColumnDefinition` to contain `options` hash

Column options are passed as an hash args then used as `options` hash in
`add_column_options!`. Converting args to attributes is inconvinient for
using options as an hash.

jeremy commented on activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb:
```diff
@@ -9,9 +9,9 @@ module ConnectionAdapters #:nodoc:
-        primary_key || type.to_sym == :primary_key
+    ColumnDefinition = Struct.new(:name, :type, :options, :sql_type) do # :nodoc:
+      def [](key)
+        options[key]
```
This poses a compatibility challenge for third party database adapters. Could we offer an gentler upgrade path? Possibly by deprecating the old methods:
```ruby
class ColumnDefinition
  attr_reader :name, :type, :sql_type, :options

  def initialize(*args, options: nil)
    if options
      initialize_with_options(*args, options)
    else
      initialize_with_struct_args(*args)
    end
  end

  def initialize_with_options(name, type, sql_type, options)
    @name, @type, @options, @sql_type = name, type, options, sql_type
  end

  def initialize_with_struct_args(name, type, limit, etc…)
    ActiveSupport::Deprecation.warn …
    @name, @type, @sql_type = name, type, sql_type
    @options = { limit: limit, precision: precision, … }
  end

  %i[ limit precision … ].each do |option_name|
    define_method(option_name) do
      ActiveSupport::Deprecation.warn …
      options[option_name]
    end

    define_method("#{option_name}=") do |value|
      ActiveSupport::Deprecation.warn …
      options[option_name] = value
    end
  end
end
```

jeremy replied:
After reading the rest of the PR, perhaps we could provide these reader/writer methods without deprecation.

jeremy commented on activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb:
```diff
@@ -305,7 +305,7 @@ def column(name, type, options = {})
         options = options.dup
 
-        if @columns_hash[name] && @columns_hash[name].primary_key?
+        if @columns_hash[name] && @columns_hash[name][:primary_key]
```
This usage feels like a regression. It's nice to call a clear predicate method to ask about the column's attributes rather than digging in to the column options to fetch the data.

jeremy commented on activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb:
```diff
@@ -360,28 +360,16 @@ def references(*args, **options)
       alias :belongs_to :references
 
-      def new_column_definition(name, type, options) # :nodoc:
+      def new_column_definition(name, type, **options) # :nodoc:
```
+keyword args for options.

jeremy commented on activerecord/lib/active_record/connection_adapters/postgresql/schema_definitions.rb:
```diff
@@ -173,24 +173,8 @@ def xml(*args, **options)
-
-          def create_column_definition(name, type)
-            PostgreSQL::ColumnDefinition.new name, type
-          end
```
Nice simplification. Pays off.

jeremy commented on activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:
```diff
@@ -0,0 +1,20 @@
+      class SchemaCreation < AbstractAdapter::SchemaCreation # :nodoc:
+        private
+          def visit_ColumnDefinition(o)
+            o.sql_type = type_to_sql(o.type, o[:limit], o[:precision], o[:scale], o[:array])
```
Seeing this repeated. Good candidate to switch to keyword args also.

jeremy commented on activerecord/lib/active_record/connection_adapters/sqlite3/schema_creation.rb:
```diff
@@ -1,15 +1,8 @@
-            options[:null] = false if o.primary_key
-            options
-          end
-
```
👍

jeremy commented on activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb:
```diff
@@ -418,21 +418,15 @@ def change_column_null(table_name, column_name, null, default = nil) #:nodoc:
-            self.precision = options[:precision] if options.include?(:precision)
-            self.scale = options[:scale] if options.include?(:scale)
-            self.collation = options[:collation] if options.include?(:collation)
+            self.options = options
```
This appears to overwrite all the existing column options instead of just the ones provided by `change_column`.

kamipo replied:
ah... good catch. I'll fix it.

jeremy commented on activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb:
```diff
@@ -418,21 +418,15 @@ def change_column_null(table_name, column_name, null, default = nil) #:nodoc:
         end
         alter_table(table_name) do |definition|
-          definition[column_name].null = null
+          definition[column_name].options[:null] = null
```
These cases of modifying a data collection directly feel like a step back from calling specific writer methods.

jeremy reviewed (commented):
I like the improvement to creating column definitions. Much simpler.

But the tradeoff is that code is accessing `column.options[…]` directly and breaking compat with other adapters.

Perhaps there's a path to using options hash and keyword args to construct column definitions, but expose an opaque attr-style API for the option values.

kamipo committed:
Restore `primary_key?` predicate method

It's nice to call a clear predicate method to ask about the column's
attributes rather than digging in to the column options to fetch the
data.

kamipo committed:
Switch to keyword args for `type_to_sql`

kamipo committed:
Define reader/writer methods for `options`

kamipo commented:
I addressed to switch to keyword args for `type_to_sql` and define reader/writer methods for `options`.

Originally `ColumnDefiniton` is a short-lived object and is only used in `create_table`, `change_table`, `change_column`, `change_column_null`, and `change_column_default`.

So it is enough to address only the changes of `type_to_sql` for third party adapters.

I created a patch for oracle-enhanced and sqlserver-adapter.

oracle-enhanced:
https://gist.github.com/kamipo/d718a823f61d13db736975fd2d7b3a5b#file-oracle-enhanced-patch

sqlserver-adapter:
https://gist.github.com/kamipo/d718a823f61d13db736975fd2d7b3a5b#file-sqlserver-adapter-patch

How about it?

jeremy commented:
Merged @ ae39b1a03d0a859be9d5342592c8936f89fcbacf

yahonda commented:
@kamipo Thanks for making a patch for Oracle enhanced adapter. Would you mind pushing the branch to github. of course pull request is welcomed.

kamipo commented:
Yeah, I'll create the branch and pull request.

davidalejandroaguilar commented on activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb:
```diff
@@ -360,28 +372,16 @@ def references(*args, **options)
-        column.as          = options[:as]
-        column
+        options[:primary_key] ||= type == :primary_key
+        options[:null] = false if options[:primary_key]
```
@kamipo @jeremy  I'm wondering why this was added? This prevents setting `null: true` when combined with the `primary_key` option. e.g.

```rb
add_reference(
  :orders,
  :user,
  null: true,
  foreign_key: true,
  primary_key: :custom_id
)
```

Will output:

```sql
ALTER TABLE "orders" ADD "user_id" uuid NOT NULL
```

Even when you specified `null: true`

Maybe I'm missing something?

davidalejandroaguilar replied:
Hey @kamipo @jeremy , I'm wondering why this got added here? Just noticed this got moved from the sqlite3 adapter into abstract. Not sure why a column could not be null if you specify the `primary_key` option.

Sorry if you get a duplicate notification, for some reason my previous comment got deleted.
